src/view/domconverter.js

```diff
@@ -338,12 +338,15 @@ export default class DomConverter {
 	domSelectionToView( domSelection ) {
 		const viewSelection = new ViewSelection();
 
+		const isBackward = this.isDomSelectionBackward( domSelection );
+
 		for ( let i = 0; i < domSelection.rangeCount; i++ ) {
+			// DOM Range have correct start and end, no matter what is the DOM Selection direction. So we don't have to fix anything.
 			const domRange = domSelection.getRangeAt( i );
 			const viewRange = this.domRangeToView( domRange );
 
 			if ( viewRange ) {
-				viewSelection.addRange( viewRange );
+				viewSelection.addRange( viewRange, isBackward );
```

Comment:
I could push all ranges to one array and then use `setRanges` but what's the point? The code is more messy and you have one more variable. When you add a range it does not matter how the other were selected, both from API point of view, DOM point of view and logical point of view.